src/mlir/design.md

```diff
@@ -0,0 +1,158 @@
+lowering while still maintaining some high level concepts to allow us to optimise
+the code without having to scan the whole IR on every pass.
+
+The LLVM pipeline intermixes analysis passes with transformation passes and has
+a way of tracking which transformation pass invalidates which analysis pass to
+avoid re-running complex analysis passes. However, more often than not, the
+transformation passes end up invalidating them anyway and a new run is needed.
+
+In Verona, we'll be bundling all modules, including the standard library and all
+source files, into one module. A full scan on that module for every analysis pass
+would be prohibitively expensive.
+
+LLVM IR has a fixed shape, with instructions belonging to basic blocks, which in
+turn belong to functions. Instructions have "uses", which we can be traversed
+cheaply, attributes and metadata, which can belong to more than one symbol, but
+that only gets us so far.
+
+Verona needs to know the type constraints of objects (imm, mut, iso) as well as
+which region they belong to and if those regions have the same sentinel object.
+We may also want to do intra-module optimisations if we know a symbol doesn't
+escape the context (which is much cheaper than whole program optimisation).
+
+For those reasons, Verona will need a light dialect to represent type constraints,
+region information and module boundaries, which don't naturally convert to LLVM.
+
+## Generation Simplicity
+
+The new AST is much simpler than the previous. The new parser uses a lean typed
+AST and converts all language into those nodes, which becomes much simpler to
+lower to an IR. The parser also runs the type checks and inference, so the AST
+we get at this stage is already complete, and checked for syntactic and semantic
+errors.
+
+Lowering errors can still occur, if the semantics would be correct in some cases
+but not in this specific case, for example, when using it with a different
+runtime functionality.
+
+The MLIR dialect that we'll develop at this stage will be solely for representing
+the lowering concepts in order to simplify optimisations. In theory, the AST
+could have been directly lowered to LLVM IR, but the LLVM pipeline wouldn't be
+able, for example, to optimise based on type and region information.
+
+## Module Boundaries
+
+Some symbols are private to the module and don't escape its boundaries. If we
+lower all functions already mangled into the IR, we'd have to scan the whole
+module every time for the functions and types that belong to them, to run module
+local optimisations.
+
+For this reason, we have chosen to create an MLIR `module` for each class/module,
+including the root module. For example:
+```
+// Verona code in directory FooBar
+class Foo { foo() { ... } }
+
+// Equivalent MLIR structure
+module @__ {
+  module @FooBar {
+    module @Foo {
+      func @foo() {
+        ...
+      }
+    }
+  }
+}
+```
+
+This structure, however, does not convert naturally to LLVM dialect. So we have
```

Comment:
I think the natural way of lowering this would be to a single LLVM module + ThinLTO module summary.  We could then run a ThinLTO-like CodeGen pipeline where each thread compiles one Verona module in parallel but also has (read-only) access to all of the others for the purpose of inlining.  If we're doing some early inlining in MLIR (especially lambda inlining for programmable control flow) then this should let us scale the compiler up to multiple cores fairly cleanly.

The mangling is still required to avoid symbol conflicts but all of the symbols are marked with private visibility or equivalent and so are erased at final link time.

Reply:
You're thinking way ahead of this example. This is the first lowering into MLIR. This representation allows us to use the symbol table of the modules to refer to functions more easily, as well to have independent modules in which to run passes in parallel (ex. `isolatedFromAbove`).

We have a pass that transforms this into a mangled linear module, which can be run at any time during the pipeline. We can even transform that into another form before the mangled sausage that needs to lower to LLVM, or we can later not even have a sausage to begin with.

The analysis/optimisation passes should pass on the representation that is most efficient, and the final representation should also allow us to do what we want wrt LTO. But all this is well after what we have now, and what we have now is so simple that won't get in the way of what we need later.